src/lib/coda_net2/coda_net2.mli

```diff
@@ -0,0 +1,255 @@
+  val random : net -> t Deferred.Or_error.t
+
+  val to_string : t -> string
+
+  val to_peerid : t -> peer_id
+end
+
+(** A "multiaddr" is libp2p's extensible encoding for network addresses.
+
+  They generally look like paths, and are read left-to-right. Each protocol
+  type defines how to decode its address format, and everything leftover is
+  encapsulated inside that protocol.
+
+  Some example multiaddrs:
+
+  - [/ipfs/QmcgpsyWgH8Y8ajJz1Cu72KnS5uo2Aa2LpzU7kinSupNKC]
+  - [/ip4/127.0.0.1/ipfs/QmcgpsyWgH8Y8ajJz1Cu72KnS5uo2Aa2LpzU7kinSupNKC/tcp/1234]
+  - [/ip6/2601:9:4f81:9700:803e:ca65:66e8:c21]
+ *)
+module Multiaddr : sig
+  type t
+
+  val to_string : t -> string
+
+  val of_string : string -> t
+end
+
+module PeerID : sig
+  type t = peer_id
+
+  val to_string : t -> string
+
+  val of_keypair : Keypair.t -> t
+end
+
+module Pubsub : sig
+  (** A subscription to a pubsub topic. *)
+  module Subscription : sig
+    type t
+
+    (** Publish a message to this pubsub topic.
+    *
+    * Returned deferred is resolved once the publish is enqueued locally.
+    * This function continues to work even if [unsubscribe t] has been called.
+    * It is exactly [Pubsub.publish] with the topic this subscription was
+    * created for, and fails in the same way. *)
+    val publish : t -> string -> unit Deferred.t
+
+    (** Unsubscribe from this topic, closing the write pipe.
+    *
+    * Returned deferred is resolved once the unsubscription is complete.
+    * This can fail if already unsubscribed. *)
+    val unsubscribe : t -> unit Deferred.Or_error.t
+  end
+
+  (** Publish a message to a topic.
+  *
+  * Returned deferred is resolved once the publish is enqueued.
+  * This can fail if signing the message failed.
+  *  *)
+  val publish : net -> topic:string -> data:string -> unit Deferred.t
+
+  (** Subscribe to a pubsub topic.
+    *
+    * Fails if already subscribed. If it succeeds, incoming messages for that
+    * topic will be written to the pipe. Returned deferred is resolved with [Ok
+    * sub] as soon as the subscription is enqueued.
+    *)
+  val subscribe :
+       net
+    -> string
+    -> ( string Envelope.Incoming.t
+       , _ Strict_pipe.buffered
+       , unit )
+       Strict_pipe.Writer.t
+    -> Subscription.t Deferred.Or_error.t
+
+  (** [register_validator net topic f] validates messages on [topic] with [f] before forwarding them.
+    *
+    * [f] will be called once per new message, and will not be called again until
+    * the deferred it returns is resolved. The helper process waits 5 seconds for
+    * the result of [f] to be reported, otherwise it considers the message invalid.
+    *
+    * If multiple validators are registered for a topic, they will all be called. If
+    * any of them return [false] the message is considered invalid.
+    *
+    * If no validators are registered for a topic, all messages will be forwarded.
+    *)
+  val register_validator :
+       net
+    -> string
```

Comment:
ISTM this should take a `Subscription.t`. Unless it's possible to gossip messages without being subscribed to the topic? This whole thing where there's a strong separation between subscribing to a topic and the decision of whether to relay the messages you got from it is a bit weird from my perspective. The decision of whether to rebroadcast is dependent on the state of e.g. the transaction pool and putting the new item into the pool at the same time as you check that state is the natural way to do it. Doing them separately might cause races actually.